blog/2022-11-03-short-circuiting.md

```diff
@@ -0,0 +1,162 @@
+to guarantee that the elements written in one thread (part of `W`) does not
+overlap with the reads (`P1` and `P2`) from any other thread.
+
+The solution we have implemented in Futhark is called short-circuiting. It works
+on one of Futhark's intermediate representations, `GPUMem`, which is a version
+of the IR with kernel operations and memory optimizations. `GPUMem` looks
+something like this[^1]:
+
+```
+let A_mem: mem = alloc (n * m * sizeof(i64))
+let A: [n][m]i64@A_mem -> 0 + {(n:m), (m:1)} = segmap (i < n) {
+    let B_mem: mem = alloc (m * sizeof(i64))
+    let B: [m]i64@B_mem -> 0 {(m:1)} = segmap (j < m) {
+        let x = i * m + j
+        in x
+    }
+    in B
+}
+```
+
+If you squint a bit, this looks like regular Futhark: The `segmap`s correspond
+to `map` with an explicit index, the arrays have size-types and so on.  But we
+also have a new type, `mem`, an extra `alloc` expression which returns a value
+of type `mem` and arrays have some extra annotations: a memory block and an
+index function. The extra annotations on arrays are used to indicate where in
+memory that array resides and how the values of that array are laid out in
+memory.
+
+The short-circuit analysis works by identifying so-called "short-circuit points"
+such as the last statement in this piece of code:
+
+```
+let A_mem: mem = alloc (n * n * sizeof(i64))
+let A: [n][m]i64@A_mem -> 0 + {(n:m), (m:1)} = ...
+let B_mem: mem = alloc (m * sizeof(i64))
+let B: [m]i64@B_mem -> 0 + {(m:1)} = ...
+let A[k] = B
+```
+
+and then trying to prove that it is safe to construct `B` directly in the memory
+of `A` instead of relying on an intermediate buffer.  In the example above, we
+call `A` the destination of the short-circuit attempt, `B` the source and `W` is
+the slice that dictates how `B` should be placed inside.  The result of a
+successful short-circuiting is that `B` is updated with the type `[m]i64@A_mem
+-> k*m + {(m:1)}`, which in turn makes the copy on the last line a no-op.
+
+Right now, you can see the effect of the short-circuiting pass when compiling
+simple programs such as:
+
+```
+def iota_one_row [n][m] (i: i64) (xs: *[n][m]i64): *[n][m]i64 =
+  xs with [i] = iota m
+```
+
+or
+
+```
+def concat_iotas (i: i64) (j: i64): []i64 =
+  concat (iota i) (iota j)
+```
+
+The resulting code has fewer allocations and redundant copies, which can result
+in increased performance. For instance, our NW benchmarks saw an impact of 32%
+increased performance on the largest dataset on A100.
+
+While verifying that the short-circuiting idea and implementation was valid, we
+offloaded some of the algebraic computations to
+[Z3](https://github.com/Z3Prover/z3), an SMT solver. Though Z3 was used to
+successfully short-circuit NW and other complex programs, it could not actually
+solve the NW overlap problem by itself. Instead, we used heuristics detailed in
+the paper to simplify the problem and reduce it to a number of simpler
```

Comment:
This is the first time you mention "the paper"; I suggest prefixing the entire post by a remark that it's a readable summary of the idea in the paper you're heading off to present at SC.

Reply:
Fixed.